### ISSUE TITLE
"received unauthorized message in CHAT_ID from 'TELEGRAM_USERNAME'"

### ISSUE THREAD

**Issue by claudioita** [scraton/node-red-contrib-telegrambot-home#6]
Hi,

Again most probably the error comes from my misconfiguration, but I am not sure. This time I am trying to use a second user.
I have the following setup in the BOT configuration
Bot-Name: MY_BOT_NAME
Token: the token of MY_BOT_NAME
Users: empty
CHat Ids: MY_CHAT_ID,OTHER_CHAT_ID

In the COMMAND NODE I have:
Bot: MY_BOT_NAME
ChatID: MY_CHAT_ID,OTHER_CHAT_ID
Command: /pump

When I use the command /pump from my telegram (my phone) it works fine, when I use the other telegram account  (on other phone) I get on node-red INFO tab
"received unauthorized message in OTHER_CHAT_ID from 'TELEGRAM_OTHER_USERNAME'"

What am I doing wrong,?

**Comment by scraton** [comment]
Hi again!

The message is indeed indicating some sort of misconfiguration. I've tested this flow locally and it is working fine (triggering the command node via two different chats). My best guess as to why you are experiencing this error would have to be that the chat IDs are not correct. Presumably OTHER_CHAT_ID is not in the list on the bot configuration.

Can you double check that the chat IDs are correct?

It could also be a matter of extraneous white space somewhere. There's not much input sanitation in this plugin for sake of simplicity, so just an extra space somewhere could be causing the problem. For example, if you have the IDs set as...

```
  1234, 1235
```

...it may fail, whereas...

```
1234,1235
```

...should work. (Note the extra whitespace at the start and between the comma.) The same goes for the authorized usernames.

Can you double check those settings and see if that advice helps?

**Comment by claudioita** [comment]
Ok, it may be multiple things. Let me explain what am I doing
I have a

COMMAND NODE in which I put
`Chat ID: 184230000, 600320000`  it's ONE blank space after the comma
`Command: /pump`


COMMAND NODE goes to a 
TELEGRAM SWITCH in which I put
`Chat ID: 184230000, 600320000`
Auto-answer: ticked
Question: MY_QUESTION

and 3 answers


In MY_BOT I have
Bot-Name: MY_BOT_NAME
Token: the token of MY_BOT_NAME
Users:   claudioita,GiulianaB  
`Chat ID: 184230000, 600320000`

With the above settings the error is gone. 
Everything works fine from my phone (telegram claudioita CHAT ID 184230000)
From the other phone (telegram GiulianaB) when I enter the command /pump, the answers go to the other phone (to claudioita phone)

IS this a bug or me not seeing something?

**Comment by scraton** [comment]
Ah ok, so this is a limitation of the current plugin. It wasn’t really designed to support full fledged multi user chat bots, though hopefully it will expand to that someday.

Currently, in order to direct the response to the appropriate chat, you’ll have to create some switch nodes that look at the chat ID property of the initiating message and then send a message to the proper chat based on that.

I started looking at supporting a way to handle this automatically, but unfortunately I can’t find a way to do it right now while preserving backwards compatibility. It’s something I’m gonna have to spend more time on in the coming weeks.

If that is too complex for your setup, you may want to take a look at [node-red-contrib-chatbot](https://github.com/guidone/node-red-contrib-chatbot) which is designed more specifically for creating chat bots. It’s more difficult to setup but is presumably going to be able to handle your use case better until I can get around to supporting it here.

I’m also open to pull requests adding support for this stuff if you’d like to get your hands dirty with JavaScript. :)

**Comment by claudioita** [comment]
Unfortunately I have zero experience programming (tried to install the chatbot node-red link, and  looks very confusing to me, imagine programming ...).

 I guess another solution is to duplicate the code, of your plugin,  changing the BOT NAME and USER and CHAT_ID. Will I be able to use the same COMMAND?
This until you, or other, have will/time to change it.

Thanks.

**Comment by scraton** [comment]
Here's an example of how you could set this up.

It uses a single command node which gets passed to a switch. The switch will inspect the origin chat ID and then send a notification to the appropriate chat.

You can add your own nodes between the command and switch, but you need to ensure it doesn't modify the `msg.payload` property. Alternatively, you can add your own nodes after the switch, but that will end up duplicating logic.

It's not pretty, but it should get the job done.

```json
[
    {
        "id": "1f2f133b.571c7d",
        "type": "tab",
        "label": "Multi-Chat Command",
        "disabled": false,
        "info": ""
    },
    {
        "id": "db7dc76c.b85a6",
        "type": "telegrambot-command",
        "z": "1f2f133b.571c7d",
        "bot": "a0e6e850.a3a028",
        "command": "/test",
        "commandType": "str",
        "commandCase": false,
        "x": 170,
        "y": 280,
        "wires": [
            [
                "d84853d1.8139"
            ]
        ]
    },
    {
        "id": "d84853d1.8139",
        "type": "switch",
        "z": "1f2f133b.571c7d",
        "name": "",
        "property": "payload.chat.id",
        "propertyType": "msg",
        "rules": [
            {
                "t": "eq",
                "v": "100000000",
                "vt": "num"
            },
            {
                "t": "eq",
                "v": "200000000",
                "vt": "num"
            }
        ],
        "checkall": "true",
        "repair": false,
        "outputs": 2,
        "x": 360,
        "y": 280,
        "wires": [
            [
                "438dce31.ee77"
            ],
            [
                "532de9c9.5556d8"
            ]
        ]
    },
    {
        "id": "438dce31.ee77",
        "type": "telegrambot-notify",
        "z": "1f2f133b.571c7d",
        "name": "",
        "bot": "a0e6e850.a3a028",
        "chatId": "100000000",
        "message": "Hello, chat A!",
        "parseMode": "",
        "x": 540,
        "y": 240,
        "wires": []
    },
    {
        "id": "532de9c9.5556d8",
        "type": "telegrambot-notify",
        "z": "1f2f133b.571c7d",
        "name": "",
        "bot": "a0e6e850.a3a028",
        "chatId": "200000000",
        "message": "Hello, chat B!",
        "parseMode": "",
        "x": 540,
        "y": 320,
        "wires": []
    },
    {
        "id": "a0e6e850.a3a028",
        "type": "telegrambot-config",
        "z": "",
        "botname": "MyTelegramBot",
        "usernames": "",
        "chatIds": "100000000,200000000",
        "pollInterval": "300"
    }
]
```

**Comment by claudioita** [comment]
Many thanks, will try this solution.

But in the first COMMAND which CHAT_ID to provide? Both like this?

[{"id":"40622e49.72a22","type":"telegrambot-switch","z":"9f069afe.141a28","name":"Telegram Switch","bot":"xxx","chatId":"100000","question":"Turn the pump ON or OFF","answers":["Yes, turn it ON","Turn it OFF","Do nothing"],"outputs":3,"autoAnswerCallback":true,"x":589.0195465087891,"y":546.0039119720459,"wires":[["5d8667a.f769398"],["28939b90.7b84c4"],["e811c62e.126158"]]},{"id":"5d8667a.f769398","type":"api-call-service","z":"9f069afe.141a28","name":"ON Pump","server":"f5c85f13.80a8a","service_domain":"homeassistant","service":"turn_on","data":"{\"entity_id\":\"input_boolean.sonoff_water_pump\"}","mergecontext":"","x":867.01953125,"y":518.0039186477661,"wires":[["cd6a5757.77c018","172276f5.8a9b59"]]},{"id":"28939b90.7b84c4","type":"api-call-service","z":"9f069afe.141a28","name":"OFF Pump","server":"f5c85f13.80a8a","service_domain":"homeassistant","service":"turn_off","data":"{\"entity_id\":\"input_boolean.sonoff_water_pump\"}","mergecontext":"","x":864.01953125,"y":567.003960609436,"wires":[["2ab0c708.5c4c58"]]},{"id":"e811c62e.126158","type":"telegrambot-notify","z":"9f069afe.141a28","name":"Doing nothing, bye","bot":"e15c8049.82fb8","chatId":"100000","message":"Doing nothing, bye","parseMode":"","x":861.01953125,"y":631.0039224624634,"wires":[]},{"id":"cc1b786.9c38e88","type":"comment","z":"9f069afe.141a28","name":"/pump command","info":"","x":142.01953125,"y":499.00392723083496,"wires":[]},{"id":"4003c24e.6d403c","type":"telegrambot-command","z":"9f069afe.141a28","bot":"e15c8049.82fb8","chatId":"1000000, 20000000","command":"/pump","commandType":"str","commandCase":false,"x":141.01956176757812,"y":548.003981590271,"wires":[["10045c0c.37c6d4"]]},{"id":"10045c0c.37c6d4","type":"switch","z":"9f069afe.141a28","name":"","property":"payload.chat.id","propertyType":"msg","rules":[{"t":"eq","v":"1000000","vt":"num"},{"t":"eq","v":"2000000","vt":"num"}],"checkall":"true","repair":false,"outputs":2,"x":331.0195617675781,"y":548.003981590271,"wires":[["40622e49.72a22"],["1e9ae99.1822116"]]},{"id":"1e9ae99.1822116","type":"telegrambot-switch","z":"9f069afe.141a28","name":"Telegram Switch","bot":"172731bf.71f29e","chatId":"2000000","question":"Turn the pump ON or OFF","answers":["Yes, turn it ON","Turn it OFF","Do nothing"],"outputs":3,"autoAnswerCallback":true,"x":590.0195236206055,"y":782.0039501190186,"wires":[["1ba7382c.103988"],["75fd06c8.0d83b8"],["5eea7408.7e14cc"]]},{"id":"1ba7382c.103988","type":"api-call-service","z":"9f069afe.141a28","name":"ON Pump","server":"f5c85f13.80a8a","service_domain":"homeassistant","service":"turn_on","data":"{\"entity_id\":\"input_boolean.sonoff_water_pump\"}","mergecontext":"","x":863.0195159912109,"y":753.0039129257202,"wires":[["648e0302.51ea4c","552e6d50.e6e7f4"]]},{"id":"75fd06c8.0d83b8","type":"api-call-service","z":"9f069afe.141a28","name":"OFF Pump","server":"f5c85f13.80a8a","service_domain":"homeassistant","service":"turn_off","data":"{\"entity_id\":\"input_boolean.sonoff_water_pump\"}","mergecontext":"","x":860.0195159912109,"y":802.0039548873901,"wires":[["545fe0d.ed9182"]]},{"id":"5eea7408.7e14cc","type":"telegrambot-notify","z":"9f069afe.141a28","name":"Doing nothing, bye","bot":"172731bf.71f29e","chatId":"20000000","message":"Doing nothing, bye","parseMode":"","x":857.0195159912109,"y":866.0039167404175,"wires":[]},{"id":"e15c8049.82fb8","type":"telegrambot-config","z":"","botname":"first_bot","usernames":"claudioita","chatIds":"1000000","pollInterval":"300"},{"id":"f5c85f13.80a8a","type":"server","z":"","name":"Home Assistant","url":"http://192.168.1.1:8123","pass":"xxxxx"},{"id":"172731bf.71f29e","type":"telegrambot-config","z":"","botname":"other_bot","usernames":"LillianB","chatIds":"20000000","pollInterval":"300"}]

**Comment by scraton** [comment]
The command node actually doesn't need any chat ID. That's a user interface bug that will be fixed in the next release. So you can provide both, or just leave the field blank, it makes no difference either way since it isn't used.

Otherwise your flow looks good! Nice job wiring that up! :)

**Comment by scraton** [comment]
Hey claudioita,

I had some time to look into this finally. I think this solution will ease the usage of multi-chat flows to avoid a bunch of extra nodes and logic. I went ahead and released it as version 0.4.1 which should be available to update via node-red.

Basically all you have to do now is wire up any notify/payload/switch node to a command node, don't specify a chat ID, and the message will automatically be routed to the chat the command came from. You should no longer need any additional switch nodes.

Here's an example flow that demonstrates how it works. If you send `/test` to the a Telegram chat, the node will reply with your name back to the same chat.

```json
[
    {
        "id": "1f2f133b.571c7d",
        "type": "tab",
        "label": "Multi-Chat Command",
        "disabled": false,
        "info": ""
    },
    {
        "id": "e4f005bc.6ef598",
        "type": "telegrambot-command",
        "z": "1f2f133b.571c7d",
        "bot": "a0e6e850.a3a028",
        "command": "/test",
        "commandType": "str",
        "commandCase": false,
        "x": 170,
        "y": 140,
        "wires": [
            [
                "50b69f65.006fb"
            ]
        ]
    },
    {
        "id": "15159402.501e64",
        "type": "telegrambot-notify",
        "z": "1f2f133b.571c7d",
        "name": "",
        "bot": "a0e6e850.a3a028",
        "chatId": "",
        "message": "",
        "parseMode": "",
        "x": 540,
        "y": 140,
        "wires": []
    },
    {
        "id": "50b69f65.006fb",
        "type": "template",
        "z": "1f2f133b.571c7d",
        "name": "",
        "field": "payload",
        "fieldType": "msg",
        "format": "handlebars",
        "syntax": "mustache",
        "template": "Hello {{telegram.from.first_name}}!",
        "output": "str",
        "x": 360,
        "y": 140,
        "wires": [
            [
                "15159402.501e64"
            ]
        ]
    },
    {
        "id": "a0e6e850.a3a028",
        "type": "telegrambot-config",
        "z": "",
        "botname": "MyTelegramBot",
        "usernames": "",
        "chatIds": "",
        "pollInterval": "300"
    }
]
```

Hope this little update helps your workflow! :)
